lib/core/src/Cardano/Pool/Performance.hs

```diff
@@ -0,0 +1,158 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- A self contained module for ranking pools according to the delegation design
+-- spec.
+--
+-- * Definitions from Delegation Design Spec:
+--
+-- \tau -> fraction of the rewards that go to the treasury
+-- R -> total rewards from one epoch that will be split among stake-holders
+-- a_0 : [0, ∞) -> leader-stake influence on pool rewards
+-- z_0 = 1/k -> size of a saturated pool
+-- σ -> relative stake of pool (includes stake from leader)
+-- s -> relative stake of pool leader
+-- σ’  = min(σ, z_0)
+-- s’ = min(s, z_0)
+-- p_j_expected = n_j / max(N_expected, 1)
+-- \hat{f} -> pool rewards (includes average!)
+-- c -> pool costs in ADA
+-- m : [0, 1]  -> margin
+-- σ : [0,1]
+--
+--
+-- TODO: rename module
+-- - Cardano.Pool.Desirability?
+-- - Cardano.Pool.Ranking?
+module Cardano.Pool.Performance
+    ( -- * Types
+      RankingConstants (..)
+    , Pool (..)
+    , RelativeStakeOf (..)
+    , Lovelace (..)
+    , Margin (..)
+    , type (>=) (..)
+
+      -- * Functions
+    , desirability
+    , unsafeToBounded
+    )
+
+    where
+
+import Data.Proxy
+    ( Proxy (..) )
+import GHC.Generics
+    ( Generic )
+import GHC.TypeLits
+    ( KnownNat, Nat, Symbol, natVal )
+import Prelude
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+newtype a >= (b :: Nat) = Bounded { unBounded :: a }
```

Comment:
Will likely remove.

Reply:
What about using something more like `newtype NonNegative a = NonNegative a` ? A phantom type sounds a bit convoluted.

Reply:
Tbh, I really like how `Double >= 0` reads.  But, yes, that seems decent. There are parse issues with `>= 0` and hlint.

So going with 2:

### 1. `Double >= 0`, `Double >= 1`
- Tbh, I do really like how this reads
- TypeOperators, Nats
- Looks like we only need the `>= 0` and `>= 1` cases, so it is more general than we need.
- Causes parse-issues with both stylish-haskell and hlint. Even with `hlint` built on `master`.

### 2. `NonNegative Double`, `Positive Double`
- Similar to 1, but less fancy

### 3. Enforce in each newtype: `Lovelace`, `TotalRewards`
- We'd need *a lot of* new types. Seems bad to me.
- `TotalRewards` are in fact in Lovelace, but would have to be separate, unrelated newtype.